cmd/srcd/cmd/root.go

```diff
@@ -88,37 +88,89 @@ func initConfig() {
 
 var logMsgRegex = regexp.MustCompile(`.*msg="(.+)"`)
 
-func logAfterTimeout(header string) chan struct{} {
-	logs, err := daemon.GetLogs()
-	if err != nil {
-		logrus.Errorf("could not get logs from server container: %v", err)
+func logAfterTimeout(msg string, timeout time.Duration) func() {
+	d := &Defered{
+		Timeout: timeout,
+		Msg:     msg,
 	}
 
-	started := make(chan struct{})
-	go func() {
-		defer logs.Close()
+	return d.Print()
+}
+
+func logAfterTimeoutWithSpinner(msg string, timeout time.Duration, spinnerInterval time.Duration) func() {
+	d := &Defered{
+		Timeout:         timeout,
+		Msg:             msg,
+		Spinner:         true,
+		SpinnerInterval: spinnerInterval,
+	}
+
+	return d.Print()
+}
+
+func logAfterTimeoutWithServerLogs(msg string, timeout time.Duration) func() {
+	logs, err := daemon.GetLogs()
+
+	inputFn := func(stop <-chan bool) <-chan string {
+		ch := make(chan string)
+
+		if err != nil {
+			logrus.Errorf("could not get logs from server container: %v", err)
+			close(ch)
+			return ch
+		}
 
-		select {
-		case <-time.After(3 * time.Second):
-			logrus.Info(header)
+		go func() {
+			defer logs.Close()
 			scanner := bufio.NewScanner(logs)
-			for scanner.Scan() {
-				select {
-				case <-started:
-					return
-				default:
-					match := logMsgRegex.FindStringSubmatch(scanner.Text())
-					if len(match) == 2 {
-						logrus.Info(match[1])
-					}
+			for msg := range scanWithTimeout(scanner, 250*time.Millisecond, stop) {
+				match := logMsgRegex.FindStringSubmatch(msg)
+				if len(match) == 2 {
+					ch <- match[1]
 				}
 			}
+
+			close(ch)
 			if err := scanner.Err(); err != nil && err != context.Canceled {
 				logrus.Errorf("can't read logs from server: %s", err)
 			}
-		case <-started:
+		}()
+
+		return ch
+	}
+
+	d := &Defered{
+		Timeout: timeout,
+		Msg:     msg,
+		InputFn: inputFn,
+	}
+
+	return d.Print()
+}
+
+func scanWithTimeout(s *bufio.Scanner, timeout time.Duration, stop <-chan bool) <-chan string {
+	texts := make(chan string)
+	c := make(chan bool)
+
+	scan := func() {
+		c <- s.Scan()
+	}
+
+	go func() {
+		go scan()
+		for {
+			select {
+			case <-stop:
+				close(texts)
+				return
+			case <-time.After(timeout):
```

Comment:
why do you need this timeout?

Reply:
Actually, the first version of the revised code wasn't like this, but it was much more similar to how we're doing [here](https://github.com/src-d/engine/blob/master/cmd/srcd/cmd/root.go#L105). But I changed it when I was testing it.

The problem was that `scanner.Scan()` is blocking so when no logs are being printed, it was just hanging and was never able to read from the `stop` channel. I don't know why and how it was working [here](https://github.com/src-d/engine/blob/master/cmd/srcd/cmd/sql.go#L55), but when I tried to debug it `scanner.Scan()` was not hanging (I don't remember if it was returning `false` though). However, the server not logging anything could actually be a real case and that's how I thought to handle it. In practice, it's a sort of polling. Do you want me to investigate this further? Or do you have any other idea?

Reply:
Ahhhhhh. I see. You need timeout indeed and second goroutine. Sorry, my bad.

Reply:
wait! but you run Scan function in goroutine also! `go scan()`
so `select` already handles stop correctly, no?

Reply:
The problem is that it hangs on `s.Scan()` so it actually never reaches the select when no logs are being printed. That's why I needed to run it in another goroutine.

Reply:
Maybe I could add a test for the `scanWithTimeout` function and test this particular case.

Reply:
it should "hang" on scan. But you do it in goroutine and write to a channel. So the `select` listens to the channel not `scan` and shouldn't be blocked.

It's a good idea to add a simple test that reproduces the problem in case I don't understand something it will help. And later on it will be much easier to figure out why the code is written like this if there is a test.

Reply:
I'm having difficulties in trying to reproduce the problem. If I write a test with a simple `bytes.Buffer`, calling `Scan()` doesn't hangs. I tried using `delve` to understand where exactly the code hangs when using `s.Scan()` on docker logs. It hangs [here](https://github.com/golang/go/blob/release-branch.go1.11/src/net/http/transport.go#L1497).

Reply:
But I can try implementing a blocking reader and use it for testing to simulate this behaviour.

Reply:
From the doc of the `Reader` interface:

> // Read reads up to len(p) bytes into p. It returns the number of bytes
// read (0 <= n <= len(p)) and any error encountered. Even if Read
// returns n < len(p), it may use all of p as scratch space during the call.
// If some data is available but not len(p) bytes, Read conventionally
// returns what is available instead of waiting for more.

But in this case, it's actually waiting. I'm gonna implement a buffer that waits on read.

Reply:
I still don't understand your problem but here is an example of blocking reader and an example of reading from it without blocking main thread without timeouts:
https://play.golang.org/p/fWgDIJhSt-6

you can change some params (like Delay) to reproduce your case.

Reply:
As discussed on slack here's the updated version https://github.com/src-d/engine/pull/216/commits/a15e8ca1514e2bd7410619c4236800f2e194c4e7.